packages/hash/frontend/src/components/hooks/use-get-workspace-shortname-by-entity-uuid.ts

```diff
@@ -0,0 +1,75 @@
+import { extractBaseUri } from "@blockprotocol/type-system";
+import { systemUserShortname } from "@hashintel/hash-shared/environment";
+import { types } from "@hashintel/hash-shared/ontology-types";
+import { entityIdFromOwnedByIdAndEntityUuid } from "@hashintel/hash-subgraph";
+import { getRootsAsEntities } from "@hashintel/hash-subgraph/src/stdlib/element/entity";
+import { useEffect, useMemo, useState } from "react";
+import { useBlockProtocolGetEntity } from "./blockProtocolFunctions/knowledge/useBlockProtocolGetEntity";
+import { useGetWorkspaceByShortname } from "./use-get-workspace-by-shortname";
+
+export const useGetWorkspaceShortnameByEntityUuid = (params: {
+  entityUuid: string;
+}): { workspaceShortname?: string; loading: boolean } => {
+  const { entityUuid } = params;
+
+  const [workspaceShortname, setWorkspaceShortname] = useState<string>();
+  const [loading, setLoading] = useState<boolean>(false);
+
+  const { getEntity } = useBlockProtocolGetEntity();
+
+  const {
+    workspace: systemUserWorkspace,
+    loading: loadingSystemUserWorkspace,
+  } = useGetWorkspaceByShortname(systemUserShortname);
+
+  const systemUserOwnedById = useMemo(
+    () =>
+      systemUserWorkspace
+        ? systemUserWorkspace.kind === "user"
+          ? systemUserWorkspace.userAccountId
+          : systemUserWorkspace.orgAccountId
+        : undefined,
+    [systemUserWorkspace],
+  );
+
+  useEffect(() => {
+    if (systemUserOwnedById) {
+      void (async () => {
+        setLoading(true);
+
+        const entityId = entityIdFromOwnedByIdAndEntityUuid(
+          systemUserOwnedById,
+          entityUuid,
+        );
+
+        const { data: subgraph } = await getEntity({
+          data: { entityId },
+        });
+
+        if (!subgraph) {
+          throw new Error(
+            `Could not get workspace entity with entity ID "${entityId}"`,
+          );
+        }
+
+        const entity = getRootsAsEntities(subgraph)[0]!;
+
+        const shortname =
+          entity.properties[
+            extractBaseUri(types.propertyType.shortName.propertyTypeId)
+          ];
+
+        if (!shortname) {
+          throw Error(
+            `Could not get shortname of workspace with entityUuid "${entityUuid}"`,
+          );
+        }
+
+        setWorkspaceShortname(shortname as string);
```

Comment:
Potential race condition here: `setWorkspaceShortname` from `useEffect` \#1 may take longer than for `useEffect` \#2 and we'll end up with old data in the state. We can fix this by adding `const cancelled = false` to the first line of `useEffect` and then set it to true in `return () => {...}`.

Reply:
Thanks for spotting! Have addressed in https://github.com/hashintel/hash/pull/1650/commits/e26406722d1dfe390e608a9f679181da90ce2d29